### PR TITLE
[bugfix](schema_change) Fix the coredump when doubly write during schema change

### PR DESCRIPTION
**Pull request by Lchangliang**
## Proposed changes

Issue Number: close #xxx

Associated with https://github.com/apache/doris/pull/18822, BE shouldn't see the prefix "_shadow_" in column name. It still has some bug when doing schema change.
During altering column from int to double, insert more values will doubly write to new/old tablets and it will coredump because the insert value doesn't match with new tablet column.
```
For example:
table schema: c1 int, c2 int
alter c2 from int to double
insert values (1, 1)
In BE, the tuple slots: c1 int, c2 int, _shadow_c2 double
new tablet schema: c1 int, c2 double.
```
Before this fix, we only use column name to match so that new tablet will read the "c2 int" slot. And then the int value doesn't match double column.
Fix it by using <column_name, type> to match slot.
```
Now for example:
table schema: c1 int, c2 int
alter c2 from int to double
insert values (1, 1)
In BE, the tuple slots: c1 int, c2 int, c2 double (remove prefix "_shadow_" too)
new tablet schema: c1 int, c2 double.
```
it will use (c2, double) to find the correct slot.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Lchangliang** [comment]
run buildall

**Review by Yukang-Lian — APPROVED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.38 seconds
 stream load tsv:          504 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162090520 Bytes

**Comment by Lchangliang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.85 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162277607 Bytes

**Comment by Lchangliang** [comment]
run feut

**Comment by Lchangliang** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.73 seconds
 stream load tsv:          515 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162114559 Bytes

**Comment by Lchangliang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.01 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162252083 Bytes

**Comment by Lchangliang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.12 seconds
 stream load tsv:          511 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162017497 Bytes

**Comment by Lchangliang** [comment]
run p0

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.92 seconds
 stream load tsv:          511 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162385505 Bytes

**Comment by Lchangliang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.49 seconds
 stream load tsv:          520 seconds loaded 74807831229 Bytes, about 137 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17161981376 Bytes

**Comment by Lchangliang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.06 seconds
 stream load tsv:          520 seconds loaded 74807831229 Bytes, about 137 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162038760 Bytes

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by Lchangliang** [comment]
run buildall

**Review by eldenmoon — DISMISSED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.86 seconds
 stream load tsv:          519 seconds loaded 74807831229 Bytes, about 137 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162051186 Bytes

**Comment by Lchangliang** [comment]
run feut

**Review by dataroaring — APPROVED** [review]
LGTM

File: be/src/exec/tablet_info.cpp [hunk 1]
```diff
@@ -66,30 +68,31 @@ Status OlapTableSchemaParam::init(const POlapTableSchemaParam& pschema) {
...
-    std::map<std::string, SlotDescriptor*> slots_map;
+    std::unordered_map<std::pair<std::string, std::string>, SlotDescriptor*> slots_map;
     _tuple_desc = _obj_pool.add(new TupleDescriptor(pschema.tuple_desc()));
 
     for (auto& p_slot_desc : pschema.slot_descs()) {
         auto slot_desc = _obj_pool.add(new SlotDescriptor(p_slot_desc));
         _tuple_desc->add_slot(slot_desc);
-        slots_map.emplace(slot_desc->col_name(), slot_desc);
+        string data_type;
+        EnumToString(TPrimitiveType, to_thrift(slot_desc->col_type()), data_type);
```
Comment (dataroaring) [inline be/src/exec/tablet_info.cpp:78]: we should check we get a type here, otherwise an unexpected behavior happens.

**Review by dataroaring — CHANGES_REQUESTED** [review]
hold, I will review it more tomorrow.

File: be/src/exec/tablet_info.cpp (hunk 1 above)
Comment (Lchangliang) [reply be/src/exec/tablet_info.cpp:78]: When we init the tablet column, the field `column_type` will change from TPrimitiveType to string by this method to FieldType by `TabletColumn::get_field_type_by_string`. And the column type in columnPB is string and it changed from TPrimitiveType.

**Comment by Lchangliang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.88 seconds
 stream load tsv:          539 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162386783 Bytes

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run feut

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**